### PR TITLE
UDF Datetime2 big dates Split/Make

### PR DESCRIPTION
**Pull request by amikish**

### CODE REVIEW

File: ydb/library/yql/core/yql_expr_type_annotation.cpp
Comment (rvu1024) [inline ydb/library/yql/core/yql_expr_type_annotation.cpp:237]: На типе могут быть одновременно флаги DateType и BitDateType? Зачем отдельная проверка на отсутствие BitDateType?

Comment (amikish) [reply ydb/library/yql/core/yql_expr_type_annotation.cpp:237]: > На типе могут быть одновременно флаги DateType и BitDateType?

Да, BitDateType навешивается отдельным битом https://github.com/ydb-platform/ydb/blob/main/ydb/library/yql/public/udf/udf_data_type.h#L207

> Зачем отдельная проверка на отсутствие BitDateType?

Проверка на BitDateType как раз нужна, т.к. широкие типы имеют оба бита `DateType | BitDateType`, а функцию неявной конвертации ресурса нужно вставлять только из узкого в широкий
Попытка неявной конвертации широкого в узкий `DateTime::MakeDate(Date32('1-1-1'))` должна приводить к ошибке

File: ydb/library/yql/udfs/common/datetime2/datetime_udf.cpp
Comment (vitstn) [inline ydb/library/yql/udfs/common/datetime2/datetime_udf.cpp:2018]: убрать tz wide функции пока не разберемся в тестами на их результат

File: ydb/library/yql/core/yql_expr_type_annotation.cpp
Comment (amikish) [inline ydb/library/yql/core/yql_expr_type_annotation.cpp:236]: добавить также неявную конвертацию ресурса TM в ресурс TM64

File: ydb/library/yql/minikql/mkql_type_ops.cpp
Comment (igormunkin) [inline ydb/library/yql/minikql/mkql_type_ops.cpp:899]: `EnrichDayInfo` can be used here.

**Review by igormunkin — CHANGES_REQUESTED** [review]
@amikish, thanks for the patch! Please, consider my comments below.

File: ydb/library/yql/udfs/common/datetime2/test_bigdates/cases/SplitMake.sql
Comment (igormunkin) [reply ydb/library/yql/udfs/common/datetime2/test_bigdates/cases/SplitMake.sql]: 5066596

File: ydb/library/yql/udfs/common/datetime2/test_bigdates/cases/SplitMake.sql
Comment (igormunkin) [reply ydb/library/yql/udfs/common/datetime2/test_bigdates/cases/SplitMake.sql]: 5066596

File: ydb/library/yql/minikql/mkql_type_ops_ut.cpp
Comment (igormunkin) [reply ydb/library/yql/minikql/mkql_type_ops_ut.cpp]: 5066596

File: ydb/library/yql/minikql/mkql_type_ops.cpp
Comment (igormunkin) [reply ydb/library/yql/minikql/mkql_type_ops.cpp]: f421956

File: ydb/library/yql/minikql/mkql_type_ops.cpp
Comment (igormunkin) [reply ydb/library/yql/minikql/mkql_type_ops.cpp]: f421956

File: ydb/library/yql/minikql/mkql_type_ops.cpp
Comment (igormunkin) [reply ydb/library/yql/minikql/mkql_type_ops.cpp:882]: Irrelevant as a result of 85f3f6d.

File: ydb/library/yql/udfs/common/datetime2/test_bigdates/ya.make
Comment (igormunkin) [inline ydb/library/yql/udfs/common/datetime2/test_bigdates/ya.make:13]: I guess `YQL_ABI_VERSION` is missing here. See other test suites for UDFs.

**Review by igormunkin — CHANGES_REQUESTED** [review]
@amikish, thanks for the fixes! There are still several issues regarding YQL ABI version (including the [one](https://github.com/ydb-platform/ydb/pull/8084/files/faa932845f827e9010243038eeed0ea14fede574#r1745213296) I've mentioned above).

File: ydb/library/yql/public/udf/udf_value_builder.h
Comment (igormunkin) [reply ydb/library/yql/public/udf/udf_value_builder.h:118]: As a result of #8914, `UDF_ABI_COMPATIBILITY_VERSION` is already 2.39.0, so 2.40.0 is required for this patchset.

File: ydb/library/yql/udfs/common/datetime2/datetime_udf.cpp
Comment (amikish) [reply ydb/library/yql/udfs/common/datetime2/datetime_udf.cpp:728]: I will better add values to exceptions where they are thrown in datetime64.h

File: ydb/library/yql/udfs/common/datetime2/test_bigdates/ya.make
Comment (amikish) [reply ydb/library/yql/udfs/common/datetime2/test_bigdates/ya.make:13]: May be you meen `ydb/library/yql/udfs/common/datetime2/ya.make` ?

File: ydb/library/yql/udfs/common/datetime2/datetime_udf.cpp
Comment (igormunkin) [reply ydb/library/yql/udfs/common/datetime2/datetime_udf.cpp:728]: I doubt, there are way more informative, than the current ones: at least nobody knows the timezones IDs, you returned to the user. What I was talking about earlier is [kinda this technique](https://github.com/ydb-platform/ydb/pull/8084/files#diff-4c008caaee33f8a256fdff5c4fb3681882b7b937906c499911014e069aa739dcR928-R942): use `storage.ToString` and yield the human-readable value, that allow to easily interpret the error message and fix the problem. Consider [the following notice](https://ydb.tech/docs/en/yql/reference/udf/list/datetime#converting-time-to-date) in the docs:
> It's worth mentioning that several TzDatetime or TzTimestamp values with a positive timezone offset cannot be cast to TzDate. Consider the example below:
> ```
> SELECT CAST(TzDatetime("1970-01-01T23:59:59,Europe/Moscow") as TzDate);
> /* Fatal: Timestamp 1970-01-01T23:59:59.000000,Europe/Moscow cannot be casted to TzDate */
> ```
> Starting from the Unix epoch, there is no valid value representing midnight on 01/01/1970 for the Europe/Moscow timezone. As a result, such a cast is impossible and fails at runtime.

Anyway, I see no reason to block this PR with this nit, so let's proceed with your change.

File: ydb/library/yql/udfs/common/datetime2/test_bigdates/ya.make
Comment (igormunkin) [reply ydb/library/yql/udfs/common/datetime2/test_bigdates/ya.make:13]: Yeah, precisely.

**Review by igormunkin — APPROVED** [review]
@amikish, thanks for the patchset! LGTM after all fixes.

File: ydb/library/yql/udfs/common/datetime2/datetime_udf.cpp
Comment (amikish) [reply ydb/library/yql/udfs/common/datetime2/datetime_udf.cpp:728]: As I understand, exceptions thrown from datetime.h and datetime64.h are some kind of internal ones, and exist mainly for debug reason in case there is some bug in make/split implementation
Values fed to DateTime::Make and DateTIme::Split functions should be already valid

**Pull request merged**
